File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/actors/buffs/Burning.java
/*
 * Pixel Dungeon
 * Copyright (C) 2012-2014  Oleg Dolya
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */
package com.watabou.pixeldungeon.actors.buffs;

import com.nyrds.Packable;
import com.nyrds.pixeldungeon.ml.R;
import com.nyrds.platform.util.StringsManager;
import com.watabou.pixeldungeon.Badges;
import com.watabou.pixeldungeon.Dungeon;
import com.watabou.pixeldungeon.ResultDescriptions;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.blobs.Blob;
import com.watabou.pixeldungeon.actors.blobs.Fire;
import com.watabou.pixeldungeon.actors.hero.Doom;
import com.watabou.pixeldungeon.actors.hero.Hero;
import com.watabou.pixeldungeon.effects.Effects;
import com.watabou.pixeldungeon.items.Item;
import com.watabou.pixeldungeon.items.rings.RingOfElements.Resistance;
import com.watabou.pixeldungeon.levels.Level;
import com.watabou.pixeldungeon.scenes.GameScene;
import com.watabou.pixeldungeon.sprites.CharSprite;
import com.watabou.pixeldungeon.ui.BuffIndicator;
import com.watabou.pixeldungeon.utils.GLog;
import com.watabou.pixeldungeon.utils.Utils;
import com.watabou.utils.Random;

public class Burning extends Buff implements Doom {

	private static final float DURATION = 8f;

	@Packable
	private float left;

	class burnItem implements ItemAction {
		public Item act(Item srcItem){
			return srcItem.burn(target.getPos());
		}
		public void carrierFx(){
			Effects.burnFX( target.getPos() );
		}
		@Override
		public String actionText(Item srcItem) {
            return Utils.format(R.string.Burning_Burns, srcItem.toString());
		}
	}
		
	@Override
	public boolean act() {
		
		if (target.isAlive()) {
			
			if (target instanceof Hero) {
				Buff.prolong( target, Light.class, TICK * 1.01f );
			}
			int bonusDamage = Dungeon.depth / 2;
			target.damage( Random.Int( 1 + bonusDamage, 5 + bonusDamage ), this );
			
			applyToCarriedItems(new burnItem());
		} else {
			detach();
		}

		final Level level = Dungeon.level;

		if (level.flammable[target.getPos()]) {
			GameScene.add( Blob.seed( target.getPos(), 4, Fire.class ) );
		}
		
		spend( TICK );
		left -= TICK;
		
		if (left <= 0 ||
			Random.Float() > (2 + (float)target.hp() / target.ht()) / 3 ||
			(level.water[target.getPos()] && !target.isFlying())) {
			
			detach();
		}
		
		return true;
	}
	
	public void reignite( Char ch ) {
		left = duration( ch );
	}
	
	@Override
	public int icon() {
		return BuffIndicator.FIRE;
	}

	public static float duration( Char ch ) {
		Resistance r = ch.buff( Resistance.class );
		return r != null ? r.durationFactor() * DURATION : DURATION;
	}

	@Override
	public void onDeath() {
		
		Badges.validateDeathFromFire();
		
		Dungeon.fail( Utils.format( ResultDescriptions.getDescription(ResultDescriptions.Reason.BURNING), Dungeon.depth ) );
        GLog.n(StringsManager.getVar(R.string.Burning_Death));
	}

	@Override
	public CharSprite.State charSpriteStatus() {
		return CharSprite.State.BURNING;
	}
}

File: scripts/buffs/Burning.lua
---
--- Burning: fire DoT, scorches carried items, seeds Fire on flammable
--- cells, doused by water; java entry points: CharUtils.ignite (attach +
--- fresh duration), Char attach-damage kind gate
--- (was actors/buffs/Burning.java)
---
local RPD  = require "scripts/lib/commonClasses"
local buff = require "scripts/lib/buff"

return buff.init{
    desc  = function ()
        return {
            icon          = 2, -- BuffIndicator.FIRE
            name          = "BurningBuff_Name",
            info          = "BurningBuff_Info",
        }
    end,

    act = function(self, buff)
        local target = buff.target
        local level = RPD.Dungeon.level

        if target:isAlive() then
            if target:getEntityKind() == "Hero" then
                RPD.Buffs.Buff:prolong(target, "Light", 1.01)
            end

            local bonusDamage = math.floor(RPD.Dungeon.depth / 2)
            target:damage(RPD.Random:Int(1 + bonusDamage, 5 + bonusDamage), buff)

            RPD.CharUtils:burnCarriedItems(buff)
        else
            buff:detach()
        end

        if level.flammable[target:getPos() + 1] then
            RPD.placeBlob(RPD.Blobs.Fire, target:getPos(), 4)
        end

        buff:spend(1) -- Actor.TICK
        self.data.left = (self.data.left or 0) - 1

        -- douse in water unless flying; else burn out or roll for a shake-off
        if level.water[target:getPos() + 1] and not target:isFlying() then
            buff:detach()
            return
        end

        if self.data.left <= 0
            or math.random() > (2 + target:hp() / target:ht()) / 3 then
            buff:detach()
        end
    end,

    -- java entry: CharUtils.ignite passes factor-scaled duration
    reignite = function(self, buff, left)
        self.data.left = left
    end,

    onHeroDeath = function(self, buff)
        RPD.Badges:validateDeathFromFire()
        RPD.Dungeon:fail(RPD.JavaUtils:format(
            RPD.ResultDescriptions:getDescription(RPD.ResultReason.BURNING), { RPD.Dungeon.depth }))
        RPD.glogn(RPD.textById("Burning_Death"))
    end,

    charSpriteStatus = function(self, buff)
        return "BURNING"
    end
}
